spec.md

```diff
@@ -3149,6 +3167,274 @@ The CO MUST implement the specified error recovery behavior when it encounters t
+There are some important things for the SP and CO to consider here:
+
+- In normal operation an SP MUST terminate the stream only after ALL metadata is transmitted, using the language specific idiom for terminating a gRPC stream source.
+  This results in the CO receiving its language specific notification idiom for the end of a gRPC stream, which provides a **definitive indication that all available metadata has been received**.
+
+- If the SP encounters an error while recovering the metadata it MUST abort transmission of the stream with its language specific error idiom.
+  This results in the CO receiving the error in its language specific idiom, which will be different from the language specific idiom for the end of a gRPC stream.
+
+- It is possible that the gRPC stream gets interrupted for arbitrary reasons beyond the control of either the SP or the CO.
+  The SP will get an error when writing to the stream and MUST abort its transmission.
+  The CO will receive an error in its language specific idiom, which will be different from the language specific idiom for the end of a gRPC stream.
+
+- In all circumstances where the CO receives an error when reading from the gRPC stream it MAY attempt to **continue** the operation by re-sending its request message but with a `starting_offset` adjusted to the NEXT byte position beyond that of any metadata already received.
+  The SP MUST always ensure that the `starting_offset` requested be considered in the computation of the data range for the first message in the returned gRPC stream, though the data range of the first message is not required to actually include the `starting_offset` if there is no applicable data between the `starting_offset` and the start of the data range returned by the first message.
+
+#### `GetMetadataAllocated`
+
+The plugin must implement this RPC if it provides the SnapshotMetadata service.
+
+```protobuf
+// The GetMetadataAllocatedRequest message is used to solicit metadata
+// on the allocated blocks of a snapshot: i.e. this identifies the
+// data ranges that have valid data as they were the target of some
+// previous write operation on the volume.
+message GetMetadataAllocatedRequest {
+  // This is the identifier of the snapshot.
+  // This field is REQUIRED.
+  string snapshot_id = 1;
+
+  // This indicates the zero based starting byte position in the volume
+  // snapshot from which the result should be computed.
+  // It is intended to be used to continue a previously interrupted
+  // call.
+  // The CO SHOULD specify this value to be the offset of the byte
+  // position immediately after the last byte of the last data range
+  // received, if continuing an interrupted operation, or zero if not.
+  // The SP MUST NOT return any data ranges for which the byte_offset
+  // plus the size_bytes is less than starting_offset.
+  // The SP MAY return data ranges with byte_offset less than
+  // starting_offset if starting_offset is less than byte_offset plus
```

Comment:
"... byte_offset less than starting_offset if starting_offset is less than byte_offset ..."

sorry, i'm still confused by this phrasing. what am i missing?

Reply:
This is language I recommended, and what I was trying to achieve is to constrain the SP's behavior in case of restarting an abnormally terminated stream. Basically I want to be very precise about what starting_offset means, so that SPs don't take liberties and start somewhere other than where the CO expects. Unfortunately, the only way to do that is to describe what mathematical relationships must hold between the values in the request and response.

Reply:
Given a requested starting_offset S, and the first block_metadata B0 of the first SP response in a stream, the SP MUST ensure that B0.byte_offset <= S < B0.byte_offset + B0.size_bytes.

.. is this what you're getting at? if so, can we represent the mathematical relationship using language like this instead of the two sentences that attempt to model this in English? I see where I went wrong reading it the first time, and I probably would have not had the same trouble w/ the same in equation form.

Reply:
Yeah I'm perfectly fine using language like that. I had worried that introducing new variables like S and B0 might be extra confusing, but this seems like a case where it will be a little confusing either way, and I do like the clarity of mathematical symbols to express what is really a mathematical relationship.

@carlbraganza do you think you can reword using the language that @jdef recommends?